crates/nargo/src/ops/test.rs

```diff
@@ -25,44 +27,98 @@ pub fn run_test<B: BlackBoxFunctionSolver>(
-                        // TODO: Improve color variations on this message
-                        message: "error: Test passed when it should have failed".to_string(),
-                    },
-                    Err(_) => TestStatus::Pass,
-                }
-            } else {
-                match circuit_execution {
-                    Ok(_) => TestStatus::Pass,
-                    Err(error) => TestStatus::Fail { message: error.to_string() },
-                }
+                return TestStatus::Fail {
+                    message: "error: Test passed when it should have failed".to_string(),
+                };
             }
+            return TestStatus::Pass;
         }
-        // Test function failed to compile
-        //
-        // Note: This could be because the compiler was able to deduce
-        // that a constraint was never satisfiable.
-        // An example of this is the program `assert(false)`
-        // In that case, we check if the test function should fail, and if so, we return `TestStatus::Pass`.
-        Err(diag) => {
-            // The test has failed compilation, but it should never fail. Report error.
-            if !test_function.should_fail() {
-                return TestStatus::CompileError(diag);
-            }
+        Err(err) => err,
+    };
 
-            // The test has failed compilation, check if it is because the program is never satisfiable.
-            // If it is never satisfiable, then this is the expected behavior.
-            let program_is_never_satisfiable =
-                diag.diagnostic.message.contains("Failed constraint");
-            if program_is_never_satisfiable {
-                return TestStatus::Pass;
-            }
+    // If we reach here, then the circuit execution failed.
+    //
+    // Check if the function should have passed
+    let test_should_have_passed = !test_function.should_fail();
+    if test_should_have_passed {
+        return TestStatus::Fail { message: circuit_execution_err.to_string() };
+    }
 
-            // The test has failed compilation, but its a compilation error. Report error
-            TestStatus::CompileError(diag)
-        }
+    check_expected_failure_message(
+        test_function,
+        circuit_execution_err.user_defined_failure_message().unwrap_or_default(),
+    )
+}
+
+fn check_expected_failure_message(test_function: TestFunction, got_error: &str) -> TestStatus {
+    // Extract the expected failure message, if there was one
+    //
+    // #[test(should_fail)] will not produce any message
+    // #[test(should_fail_with = "reason")] will produce a message
+    //
+    let expected_failure_message = match test_function.failure_reason() {
+        Some(reason) => reason,
+        None => return TestStatus::Pass,
+    };
+
+    let failure_message_in_assert_method = got_error;
+    let expected_failure_message_matches =
+        failure_message_in_assert_method == expected_failure_message;
```

Comment:
```suggestion
    let expected_failure_message_matches =
        got_error == expected_failure_message;
```
Why not just this? Or did you want to rename it on purpose